PCIe controller support

rondoval opened the pull request:
This enables PCIe controller support:

- Enables 36-bit IPS on MMU; this is because the BAR window is mapped at phys address 0x600000000, so without this change the MMU page faults with DFSC=0x2 (Address size fault at level 2)
- Maps the /scb ranges in a similar fashion to how /soc was already mapped. This results with the same phys address range being mapped twice into 0xf2 area, but is cleaner - i.e. the genet driver will also be able to map genet regs via /scb tree as per spec (it currently needs to "jump" to /soc to find the mapping)
- Clamps the PCIe MMIO (BAR) to 64MiB (it is 1GiB in the devtree)
- Maps the PCIe MMIO (BAR) window into the 0xF2 area
- There's also a change in dt_find_node which enables it to find 2nd level nodes - this is used to find the PCIe node at SCB
- There are also extra properties added to the device tree, because the PCIe driver needs both the physical and virtual addresses of the MMIO window in order to set up the controller

Note: Not tested on PI3, I am not sure if the IPS change is OK for PI3. The PCIe stuff should skip once it doesn't find a PCIe node in the devtree.

michalsc commented:
Small note - you do not need to squeeze the PCIe window to 64MB. If you need and want, feel free to use another area, e.g. 0xF4xxxxxx for that, same as local ARM MMIO is mapped at 0xF3xxxxxx.

rondoval commented:
well, the VL805 needs 4KB apparently, so 64MB is over the top, unless someone is planning to unsolder the VL805 and add something more interesting :)
It uses the same allocation scheme, so effectively the window is already at 0xFAxxxxxx, just above the /soc and /scb mappings

michalsc commented:
Scheduled for merge after 1.1 release

michalsc commented:
can you rework the PR so that it can go into 1_1_wip branch instead of master?

rondoval commented:
sure, will do soon

rondoval committed:
Map ranges from /scb and PCIe BAR window.
Add phys->virt mapping for PCIe BAR window.
Switch to 36bit IPS (PCIe window is at 0x600000000).

rondoval commented:
This is now rebaselined on 1.1.

Thing is, overall, it doesn't work. I.e. these changes seem OK, both genet and XHCI are able to initialize. But both cause Emu68 to... er, hang? E.g. with genet that's easy to reproduce by using AmiSpeedTest. When it happens, the Housekeeper keeps working but is unable to reboot the board. It tires, Amiga board reboots, but Pi doesn't.

Since I have no idea yet what is going on - keeping this in draft.

===
Update: got it working; I'll amend PR with a fix in the next few days.

rondoval committed:
fix for spurious interrupt death spiral

rondoval committed:
disable CPU0 GIC controller on RESET opcode.
correct PM_RSTS partition mask on Pi reset.

rondoval commented:
OK, so I ended up adding a few changes on top of the original contents:
- added disabling GIC-400 controller on CPU 0 on 68k RESET opcode. This is so that OS doesn't get stuck on boot if devices were not shut down cleanly
- added clearing INTF.ARM if ARMPending and interrupts are being disabled. This is to avoid a race condition where INTF.ARM is raised by the int vector while Exec is disabling interrupts. This results with an endless loop where Exec discards IRQ as spurious and Emu68 raises it again.
- changed PM_RSTS mask - I believe it was reverse, i.e. keeping partition number and zeroing everything else.

Let me know if you want any of this out.

BTW: the embedded gateware "somehow" makes my Amiga stuck early in Housekeeper triggered reboot. No idea why yet (I belive the sources of this version are not available?). This happens early, i.e. in platform_init() or earlier I think, also on clean 1.1 branch. Doesn't happen with custom built gateware.

michalsc commented:
Merging. However I consider to revert the PM_RSTS - as you may see one line below I was force-setting partition to 0 on purpose. But we will see if people report any issues.
